### PR TITLE
Agent unit tests completed

### PR DESCRIPTION
**Pull request by davidohana**
GlobalNet support is still missing from controller and tests (commented out)

### CODE REVIEW

File: pkg/agent/controller/agent.go [hunk 1]
```diff
@@ -81,6 +81,9 @@ func New(spec *AgentSpecification, syncerConf broker.SyncerConfig, kubeClientSet
 		return nil, errors.Wrap(err, "error converting resource")
 	}
 
+	waitForCacheSyncLocal := true
```
Comment (elevran) [inline pkg/agent/controller/agent.go:84]: Confirming:
this settings are going to apply to all scenarios?
If the broker is not set to `true` I would expect that the default value would suffice unless you want to make this explicit (in which case `waitForCacheSyncBroker` and `waitForCacheSyncLocal` suggest the same behavior and you may want to call them, e.g.,  `waitForCacheSyncLocal` and `dontWaitForCacheSyncBroker`)

File: pkg/agent/controller/agent.go [hunk 2]
```diff
@@ -219,34 +228,42 @@ func (a *Controller) Start(stopCh <-chan struct{}) error {
 	defer utilruntime.HandleCrash()
 
 	// Start the informer factories to begin populating the informer caches
-	logger.Info("Starting Agent controller")
+	agentLogger := logger.WithValues("ClusterID", a.clusterID)
+	agentLogger.Info("Starting Agent controller")
 
+	agentLogger.V(log.DEBUG).Info("Starting Agent controller")
```
Comment (elevran) [inline pkg/agent/controller/agent.go:234]: if verbosity==DEBUG won't this Info message be printed twice? Once in normal Info (line 232) and the other in line 234?

File: pkg/agent/controller/agent.go [hunk 3]
```diff
@@ -259,6 +276,7 @@ func (a *Controller) Start(stopCh <-chan struct{}) error {
 
 	// check on startup if a local service still exist for all remote service exports uploaded from this cluster.
 	// if not - enqueue deletion of the service, to delete obsolete service export from the broker
+	agentLogger.V(log.DEBUG).Info("Checking for service-less ServiceExports on the broker")
```
Comment (elevran) [inline pkg/agent/controller/agent.go:288]: checking for obsolete...?

File: pkg/agent/controller/agent.go (hunk 2 above)
Comment (elevran) [inline pkg/agent/controller/agent.go]: if verbosity==DEBUG won't this Info message be printed twice? Once in normal Info (line 232) and the other in line 234?

File: pkg/agent/controller/agent.go [hunk 4]
```diff
@@ -81,6 +81,9 @@ func New(spec *AgentSpecification, syncerConf broker.SyncerConfig, kubeClientSet
 		return nil, errors.Wrap(err, "error converting resource")
 	}
 
+	waitForCacheSyncLocal := true
+	waitForCacheSyncBroker := false
```
Comment (elevran) [inline pkg/agent/controller/agent.go:86]: Confirming:
this settings are going to apply to all scenarios?

Also, if the broker is never set to true I would expect that the default value would suffice.
If you want to make this explicit, consider that `waitForCacheSyncBroker` and `waitForCacheSyncLocal` suggest (to me) the same behavior and you may want to call them, e.g., `waitForCacheSyncLocal` and `dontWaitForCacheSyncBroker`.

File: pkg/agent/controller/cleanup.go [hunk 5]
```diff
@@ -32,9 +32,15 @@ import (
 )
 
 var (
+	serviceExportGVR = schema.GroupVersionResource{
+		Group:    mcsv1a1.GroupName,
+		Version:  mcsv1a1.GroupVersion.Version,
+		Resource: "serviceexports",
```
Comment (elevran) [inline pkg/agent/controller/cleanup.go]: question out of ignorance - are these resource names specified in the Spec and we can just import them? The use of strings is susceptible to subtle copy-paste errors or typos

File: pkg/agent/controller/controller_suite_test.go [hunk 6]
```diff
@@ -25,9 +25,12 @@ import (
 	"fmt"
 	"github.com/submariner-io/admiral/pkg/log"
 	"github.com/submariner-io/admiral/pkg/log/kzerolog"
+	"github.com/submariner-io/admiral/pkg/resource"
```
Comment (elevran) [inline pkg/agent/controller/controller_suite_test.go]: go imports should be run

File: pkg/agent/controller/endpoint.go [hunk 7]
```diff
@@ -197,18 +193,14 @@ func (e *EndpointController) endpointSliceFromEndpoints(endpoints *corev1.Endpoi
...
-	}
+	logger.V(log.TRACE).Info("Returning EndpointSlice:\n" + PrettyPrint(endpointSlice))
 
 	return endpointSlice, false
 }
 
 func (e *EndpointController) getEndpointsFromAddresses(addresses []corev1.EndpointAddress, addressType discovery.AddressType,
 	ready bool) ([]discovery.Endpoint, bool) {
-	endpoints := []discovery.Endpoint{}
+	var endpoints []discovery.Endpoint
```
Comment (elevran) [inline pkg/agent/controller/endpoint.go]: previous line initialized an empty array, new line initializes a nil array. Difference?

File: pkg/lhutil/lighthouse.go [hunk 8]
```diff
@@ -20,6 +20,7 @@ package lhutil
 
 import (
 	"fmt"
+	corev1 "k8s.io/api/core/v1"
 	mcsv1a1 "sigs.k8s.io/mcs-api/pkg/apis/v1alpha1"
```
Comment (elevran) [inline pkg/lhutil/lighthouse.go:25]: probably bad left over from my commits but should run go imports to fix empty line and order

File: pkg/lhutil/lighthouse.go [hunk 9]
```diff
@@ -100,3 +101,14 @@ func GetServiceExportCondition(status *mcsv1a1.ServiceExportStatus, ct mcsv1a1.S
 
 	return latestCond
 }
+
+func CreateServiceExportCondition(ct mcsv1a1.ServiceExportConditionType, cs corev1.ConditionStatus, reason, msg string) *mcsv1a1.ServiceExportCondition {
+	now := metav1.Now()
```
Comment (elevran) [inline pkg/lhutil/lighthouse.go:106]: I think the `LastTransitionTime` should be set by the API masters, not by the caller.
Not sure...

File: pkg/lhutil/lighthouse.go [hunk 10]
```diff
@@ -100,3 +101,14 @@ func GetServiceExportCondition(status *mcsv1a1.ServiceExportStatus, ct mcsv1a1.S
...
+	return &mcsv1a1.ServiceExportCondition{
```
Comment (elevran) [inline pkg/lhutil/lighthouse.go:107]: I see you're warming up to eliminating intermediate variables ;-)

**Review by elevran — COMMENTED** [review]
several minor questions in comments, otherwise lgtm. Let's discuss/resolve and then we can approve and merge

File: pkg/agent/controller/agent.go (hunk 4 above)
Comment (davidohana) [reply pkg/agent/controller/agent.go:86]: This is just a simple knob to turn wait for cache sync on / off for all syncers. waiting for cache sync only for local syncers is the same behavior that was implemented in LH agent before.

File: pkg/agent/controller/agent.go (hunk 2 above)
Comment (davidohana) [reply pkg/agent/controller/agent.go]: sure, typo, will fix.

Comment (davidohana) [reply pkg/agent/controller/agent.go]: should be "Starting ServiceExport uploader"

File: pkg/agent/controller/agent.go (hunk 3 above)
Comment (davidohana) [reply pkg/agent/controller/agent.go:288]: obsolete not descriptive enough - see line 266 - those are also obsolete from another reason

File: pkg/agent/controller/cleanup.go (hunk 5 above)
Comment (davidohana) [reply pkg/agent/controller/cleanup.go]: thats the way it was implemented before.  Changed to: 
```go
func getPluralTypeName(obj interface{}) string {
	return strings.ToLower(reflect.TypeOf(obj).Name()) + "s"
}
```
Do you know of a better way?

File: pkg/agent/controller/endpoint.go (hunk 7 above)
Comment (davidohana) [reply pkg/agent/controller/endpoint.go]: none, just better practice  (quick fix suggested by GoLand)
https://stackoverflow.com/questions/45317074/best-practices-constructing-an-empty-array

File: pkg/lhutil/lighthouse.go (hunk 8 above)
Comment (davidohana) [reply pkg/lhutil/lighthouse.go:25]: yes, i will make linter happy now and push again.

File: pkg/lhutil/lighthouse.go (hunk 9 above)
Comment (davidohana) [reply pkg/lhutil/lighthouse.go:106]: LastTransitionTime is just an application field in a custom type, dont think that there is any special handling for it

File: pkg/agent/controller/cleanup.go (hunk 5 above)
Comment (elevran) [reply pkg/agent/controller/cleanup.go]: I do not. The new method seems better

**Pull request merged**